### ISSUE TITLE
Cannot apply static IP address to Raspberry Pi 4

### ISSUE THREAD

**Issue by jmgilman** [flatcar/Flatcar#431]
## Description

When attempting to configure an interface using networkd via a container linux configuration, the interface fails to come up during boot. Manually restarting the networkd service after boot resolves the issue and configures the interface correctly. 

## Impact

Cannot configure the network on a Raspberry Pi 4. 

## Environment and steps to reproduce

1. **Set-up**: Use the below configuration excerpt:
  ```yaml
  networkd:
  units:
    - name: static.network
      contents: |
        [Match]
        Name=en*

        [Network]
        Address=192.168.1.102/24
        Gateway=192.168.1.1
        DNS=1.1.1.1
  ```
2. **Task**: Configure networking
3. **Action(s)**:
  a. Boot the system using an Ignition configuration containing the above-compiled excerpt
  b. Check the status of the network interface (will not be configured)
4. **Error**:
  ```
core@localhost ~ $ sudo journalctl -u systemd-networkd
-- Journal begins at Wed 2021-06-16 16:34:50 UTC, ends at Mon 2021-07-19 21:51:55 UTC. --
  Jun 16 16:34:51 localhost systemd[1]: Starting Network Service... 
  Jun 16 16:34:51 localhost systemd-networkd[383]: Enumeration completed
  Jun 16 16:34:51 localhost systemd[1]: Started Network Service.
  Jun 16 16:34:52 localhost systemd-networkd[383]: eth0: Interface name change detected, eth0 has been renamed to enabcm6e4ei0.
  Jun 16 16:34:52 localhost systemd-networkd[383]: enabcm6e4ei0: Could not bring up interface: No such device
  Jun 16 16:34:57 localhost systemd[1]: Stopping Network Service...
  Jun 16 16:34:57 localhost systemd[1]: systemd-networkd.service: Succeeded.
  Jun 16 16:34:57 localhost systemd[1]: Stopped Network Service.
  Jun 16 16:35:06 localhost systemd[1]: Starting Network Service...
  Jun 16 16:35:06 localhost systemd-networkd[830]: Enumeration completed
  Jun 16 16:35:06 localhost systemd[1]: Started Network Service.
  Jun 16 16:35:07 localhost systemd-networkd[830]: enabcm6e4ei0: Could not bring up interface: No such device
  ```

## Expected behavior

Expected the interface to be configured with a static IP address.

## Additional information

This is what the logs look like after manually restarting networkd:

```
core@localhost ~ $ sudo journalctl -u systemd-networkd
-- Journal begins at Wed 2021-06-16 16:34:50 UTC, ends at Mon 2021-07-19 21:51:55 UTC. --
Jun 16 16:34:51 localhost systemd[1]: Starting Network Service... 
Jun 16 16:34:51 localhost systemd-networkd[383]: Enumeration completed
Jun 16 16:34:51 localhost systemd[1]: Started Network Service.
Jun 16 16:34:52 localhost systemd-networkd[383]: eth0: Interface name change detected, eth0 has been renamed to enabcm6e4ei0.
Jun 16 16:34:52 localhost systemd-networkd[383]: enabcm6e4ei0: Could not bring up interface: No such device
Jun 16 16:34:57 localhost systemd[1]: Stopping Network Service...
Jun 16 16:34:57 localhost systemd[1]: systemd-networkd.service: Succeeded.
Jun 16 16:34:57 localhost systemd[1]: Stopped Network Service.
Jun 16 16:35:06 localhost systemd[1]: Starting Network Service...
Jun 16 16:35:06 localhost systemd-networkd[830]: Enumeration completed
Jun 16 16:35:06 localhost systemd[1]: Started Network Service.
Jun 16 16:35:07 localhost systemd-networkd[830]: enabcm6e4ei0: Could not bring up interface: No such device
Jun 16 16:37:42 localhost systemd[1]: Stopping Network Service...
Jun 16 16:37:42 localhost systemd[1]: systemd-networkd.service: Succeeded.
Jun 16 16:37:42 localhost systemd[1]: Stopped Network Service.
Jun 16 16:37:42 localhost systemd[1]: Starting Network Service...
Jun 16 16:37:43 localhost systemd-networkd[993]: Enumeration completed
Jun 16 16:37:43 localhost systemd[1]: Started Network Service.
Jun 16 16:37:43 localhost systemd-networkd[993]: enabcm6e4ei0: Link UP 
Jun 16 16:37:47 localhost systemd-networkd[993]: enabcm6e4ei0: Gained carrier
Jun 16 16:37:48 localhost systemd-networkd[993]: enabcm6e4ei0: Gained IPv6LL
```

I'm not sure exactly what is happening between being booted and getting a console in userspace, but I can immediately reboot the service after booting and it will bring up the interface as expected. I see at some point it's renamed from eth0, but the match should work for the subsequent starts of the service (when it's named enabcm6e4ei0). I also tried changing the match to `eth0` and it made no difference. 

[This issue](https://github.com/raspberrypi/linux/issues/3108) may or may not be related. In particular, [this comment](https://github.com/raspberrypi/linux/issues/3108#issuecomment-723550749) talks about starting the network during the initramfs stage, which I believe flatcar is doing.

**Comment by jmgilman** [comment]
I'm seeing the below lines in `dmesg`:

```
[   20.247965] Unable to find PHY
[   20.253695] bcmgenet BCM6E4E:00 enabcm6e4ei0: failed to connect to PHY
```

This was seen in [this issue](https://github.com/systemd/systemd/issues/17509) which points back to the one linked in the original submission. I suspect this may require a change to the way drivers are being loaded but my knowledge is limited in this area.

**Comment by jmgilman** [comment]
Was able to resolve the issue by loading the following kernel modules earlier in the boot process:

```yaml
storage:
  files:
    - path: /etc/modules-load.d/net.conf
      filesystem: root
      mode: 0644
      contents:
        inline: | 
          bcm_phy_lib 
          broadcom 
          mdio_bcm_unimac 
          genet
```

Since this is pretty specific to the hardware and can be resolved with a configuration change, I don't think this qualifies as a bug. However, hopefully, this thread helps a future RPi4 user.

**Comment by jmgilman** [comment]
Sorry for the flurry of activity, I am actively working on this today and have been jumping over a lot of hurdles. I'm opening this back up because even though the above fix resolved the issue of the machine not coming up with a static IP address, it doesn't appear to solve all boot-time networking-related issues.

In particular, attempting to download remote files via Ignition causes Ignition to fail. Unfortunately I have another bug that prevents any console output to my monitor after the EFI boot, so I can't see exactly what the error is, but adding this tidbit of configuration causes it to not boot:

```
storage:
  files:
    - path: /root/.docker/cli-plugins/docker-compose
      filesystem: root
      mode: 0755
      contents:
        remote:
          url: https://github.com/docker/compose-cli/releases/download/v2.0.0-beta.6/docker-compose-linux-arm64
          verification:
            hash:
              function: sha512
              sum: 7d2d5eaea7fc7caf97123330a2196f1591cc6553eea1490a6353444119f3da023e6224cae4cfb14d2ea43473a028363841e5e05b325b06e4f5d7bb28f0b47d1a
```

Removing it allows the system to boot normally, so I'm assuming there's something going wrong in Ignition (I haven't worked out how to get to the logs when it won't boot). I tested the above config in QEMU and it downloaded the file as expected.

I noticed that, whereas I was previously seeing the PHY failure a few times in `dmesg`, I am now only seeing it once early in the boot. I suspect my fix solved the latter cases but not this one and it's likely still causing issues with Ignition.

[The full dmesg can be seen here.](https://gist.github.com/jmgilman/51bdd5c3c04275456e0896ed9a3c8b01)

If this is better suited for a new issue let me know and I can open up a separate one - but it seems to be correlated to the same problem of not having networking during the early boot phase. After browsing other issues, [this one may be related](https://github.com/kinvolk/Flatcar/issues/375).
